app/src/main/kotlin/com/mapbox/vision/teaser/MainActivity.kt

```diff
@@ -1,55 +1,531 @@
+    private var lastSpeed: Float = 0f
+    private var calibrationProgress = 0f
+
+    private val visionEventsListener = object : VisionEventsListener {
+
+        override fun onCountryUpdated(country: Country) {
+            this@MainActivity.country = country
+        }
+
+        override fun onFrameSignClassificationsUpdated(frameSignClassifications: FrameSignClassifications) {
+            if (visionMode == VisionMode.Classification) {
+                runOnUiThread {
+                    tracker.update(UiSign.getUiSigns(frameSignClassifications))
+                    drawSigns(tracker.getCurrent())
+                }
+            }
+        }
+
+        override fun onRoadDescriptionUpdated(roadDescription: RoadDescription) {
+            if (visionMode == VisionMode.Lanes) {
+                runOnUiThread {
+                    drawLanesDetection(roadDescription)
                 }
             }
         }
 
-        root.findViewById<LinearLayout>(R.id.replay_mode_button_container).apply {
-            setOnClickListener { showReplayModeFragment() }
+        override fun onVehicleStateUpdated(vehicleState: VehicleState) {
+            lastSpeed = vehicleState.speed
+        }
+
+        override fun onCameraUpdated(camera: com.mapbox.vision.mobile.core.models.Camera) {
+            calibrationProgress = camera.calibrationProgress
+            runOnUiThread {
+                fps_performance_view.setCalibrationProgress(calibrationProgress)
+            }
+        }
+
+        @SuppressLint("SetTextI18n")
+        override fun onUpdateCompleted() {
+            runOnUiThread {
+                if (visionManagerWasInit) {
+                    val frameStatistics = when (visionManagerMode) {
+                        Camera -> VisionManager.getFrameStatistics()
+                        Replay -> VisionReplayManager.getFrameStatistics()
+                    }
+                    fps_performance_view.showInfo(frameStatistics)
+                }
+            }
         }
     }
 
-    override fun initVisionManager(visionView: VisionView): Boolean {
-        return when (mode) {
-            Camera -> initVisionManagerCamera(visionView)
-            Replay -> initVisionManagerReplay(visionView, sessionPath)
+    private val visionSafetyListener = object : VisionSafetyListener {
+
+        private var currentDangerLevel: CollisionDangerLevel = CollisionDangerLevel.None
+
+        private val distanceFormatter by lazy {
+            LocaleUtils().let { localeUtils ->
+                val language = localeUtils.inferDeviceLanguage(this@MainActivity)
+                val unitType = localeUtils.getUnitTypeForDeviceLocale(this@MainActivity)
+                val roundingIncrement = NavigationConstants.ROUNDING_INCREMENT_FIVE
+                DistanceFormatter(this@MainActivity, language, unitType, roundingIncrement)
+            }
+        }
+
+        override fun onCollisionsUpdated(collisions: Array<CollisionObject>) {
```

Comment:
how about dividing function into smaller ones?
50 lines

Reply:
Added to issue https://github.com/mapbox/mapbox-vision-android-teaser/issues/191